File: C++ - SFML/Asteroids/src/Ship.cpp
#include "Astro.h"
#include "Ship.h"
#include "Saucer.h"
#include "Explosion.h"
#include "HighScores.h"

Ship::Ship(Astro& astro, const Animation& anim) :
    Entity(astro, anim, astro.WIDTH * 0.5f, astro.HEIGHT * 0.5f, 0),
    thrusting(false),
    extraShips(3) {
    extraShip = astro.animations["ship"].sprite;
    extraShip.setTextureRect(astro.animations["ship"].frames[0]);
    extraShip.setRotation(0);
    extraShip.setScale(0.65f, 0.65f);
}

void Ship::update() {
    if (thrusting) {
        anim = astro.animations["ship_thrusting"];
        if (astro.sounds["thrust"].getStatus() == sf::Sound::Stopped)
            astro.sounds["thrust"].play(); // play when thrusting, but not if it's already playing
        dx += cos(heading * DEG2RAD) * 0.2f;
        dy += sin(heading * DEG2RAD) * 0.2f;
    }
    else {
        anim = astro.animations["ship"];
        astro.sounds["thrust"].stop();
        dx *= 0.99f;
        dy *= 0.99f;
    }

    float speed = sqrt(dx * dx + dy * dy);
    if (speed > MAXSPEED) {
        dx *= MAXSPEED / speed;
        dy *= MAXSPEED / speed;
    }

    Entity::update();
    Entity::wrap();
}

void Ship::check_collisions() {
    for (Entity* e : astro.currentEntities) {
        if (e->active && e->collides_with(this)) {
            Explosion::spawn(astro, astro.animations["explosion_ship"], x, y);
            astro.sounds["bang_lg"].play();
            x = astro.WIDTH * 0.5f; // reset the ship to the center of the screen
            y = astro.HEIGHT * 0.5f;
            dx = dy = 0;

            if (--extraShips < 0) {
                HighScores::scores = "Loading...";
                astro.saucer->hide();
                astro.gameState = (astro.score > HighScores::lowestHighScore) ?
                    GameState::GAMEOVER : // prompt for player's name
                    GameState::READY; // show high scores
            }
        }
    }
}

void Ship::draw_extra_ships() {
    for (int i = 0; i < extraShips; i++) {
        extraShip.setPosition(1000 + 40.0f * i, 60.0f);
        astro.window.draw(extraShip);
    }
}

File: TypeScript - p5/src/Ship.ts
import P5 from "p5";
import Astro, { GameState } from "./Astro";
import Animation from "./Animation";
import Entity from "./Entity";
import Explosion from "./Explosion";
import HighScores from "./HighScores";

export default class Ship extends Entity {
  thrusting: boolean;
  extraShips: number;

  constructor(anim: Animation, x: number, y: number, heading: number = 0) {
    super(anim, x, y, heading);
    this.thrusting = false;
    this.extraShips = 3;
  }

  update() {
    if (this.thrusting) {
      this.anim = Astro.animations["ship_thrusting"];
      Astro.sounds["thrust"].play();
      this.dx += Math.cos(this.heading) * 0.2;
      this.dy += Math.sin(this.heading) * 0.2;
    } else {
      this.anim = Astro.animations["ship"];
      Astro.sounds["thrust"].stop();
      this.dx *= 0.99;
      this.dy *= 0.99;
    }

    let speed = Math.sqrt(this.dx * this.dx + this.dy * this.dy);
    if (speed > Astro.MAXSPEED) {
      this.dx *= Astro.MAXSPEED / speed;
      this.dy *= Astro.MAXSPEED / speed;
    }

    super.update();
    super.wrap();
  }

  checkCollisions() {
    for (let e of Astro.currentEntities) {
      if (e.collision(this)) {
        Astro.entities.push(new Explosion(Astro.animations["explosion_ship"], this.x, this.y));
        Astro.sounds["bang_lg"].play();
        this.x = Astro.WIDTH * 0.5; // reset the ship to the center of the screen
        this.y = Astro.HEIGHT * 0.5;
        this.dx = this.dy = 0;

        if (--this.extraShips < 0) {
          this.active = false;
          if (Astro.saucer) Astro.saucer.destroy();
          Astro.gameState =
            Astro.score > HighScores.lowest
              ? GameState.GAMEOVER // prompt for player's name
              : GameState.READY; // show high scores
        }
      }
    }
  }

  drawExtraShips(p5: P5) {
    p5.push();
    let f = Astro.animations["ship"].frames[0];
    p5.scale(0.65);
    p5.translate(1600, 100); // set the initial drawing position
    for (let i = 0; i < this.extraShips; i++) {
      p5.translate(60, 0); // draw each ship to the left of the previous one
      p5.image(Astro.animations["ship"].texture, 0, 0, f.w, f.h, f.x, f.y, f.w, f.h); // source of the image from within the larger texture
    }
    p5.pop();
  }
}
